File: my-folder/0210-course-schedule-ii/solution.py
from collections import defaultdict
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        adj = defaultdict(list)
        indegree = [0] * numCourses
        for crs, pre in prerequisites:
            adj[pre].append(crs)
            indegree[crs] += 1
        course_order = []
       
        def dfs(node):
            course_order.append(node)
            indegree[node] -= 1

            for nei in adj[node]:
                indegree[nei] -= 1
                if( indegree[nei] == 0):
                    dfs(nei)

        for i in range(numCourses):
            if(indegree[i] == 0):
                dfs(i)
            
        return course_order if len(course_order) == numCourses else []

        

        # if numCourses == 1:
        #     return [0]
        # sorted_prereqs = sorted(prerequisites, key = lambda x:(x[1],x[0]))
        # print(sorted_prereqs)
        # res= set()
        # order=[]
        # for crs,preq in sorted_prereqs:
        #     if preq not in res:
        #         order.append(preq)
        #         res.add(preq)
        #     if crs not in res:
        #         order.append(crs)
        #         res.add(crs)
        
        # return order

File: my-folder/0210-course-schedule-ii/solution.java
class Solution {
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        List<List<Integer>> adjList = new ArrayList<>();
        int[] indegree = new int[numCourses];

        for(int c = 0; c < numCourses; c++){
            adjList.add(new ArrayList<>());
        }

        for(int[] pre : prerequisites){
            indegree[pre[0]]++;
            adjList.get(pre[1]).add(pre[0]);
        }
        ArrayList<Integer> res = new ArrayList<>();
        Queue<Integer> que = new LinkedList<>();
        for(int c = 0; c < numCourses; c++){
            if(indegree[c] == 0){
                que.offer(c);
            }
        }

        while(!que.isEmpty()){
            int curr = que.poll();
            res.add(curr);
            for(int c : adjList.get(curr)){
                indegree[c]--;
                if(indegree[c] == 0){
                    que.offer(c);
                }
            }
        }

        if(res.size() != numCourses){
            return new int[0];
        }

        int[] order = new int[numCourses];
        for (int i = 0; i < numCourses; i++) {
            order[i] = res.get(i);
        }

        return order;
    }
}

// numCourses = 4, prerequisites = [[1,0],[2,0],[3,1],[3,2]]
// 1 : 0
// 2 : 0
// 3 : [1,2]

